src/kdf.rs

```diff
@@ -355,8 +365,315 @@ impl MechOperation for Sp800Operation {
+        let salt = match params.ulSaltType {
+            CKF_HKDF_SALT_NULL => {
+                if params.ulSaltLen > 0 || params.pSalt != std::ptr::null_mut()
+                {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                } else {
+                    vec![0u8; hashlen]
+                }
+            }
+            CKF_HKDF_SALT_DATA => {
+                if params.ulSaltLen == 0 || params.pSalt == std::ptr::null_mut()
+                {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                } else {
+                    bytes_to_slice!(params.pSalt, params.ulSaltLen, u8).to_vec()
+                }
+            }
+            CKF_HKDF_SALT_KEY => {
+                /* a len of 0 indicates a key object is needed.
+                 * the salt must be set via [requires/receives]_objects()
+                 * if not derive() will error */
+                Vec::new()
+            }
+            _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+        };
+
+        Ok(HKDFOperation {
+            finalized: false,
+            extract: params.bExtract != CK_FALSE,
+            expand: params.bExpand != CK_FALSE,
+            prf: params.prfHashMechanism,
+            prflen: hashlen,
+            salt_type: params.ulSaltType,
+            salt_key: [params.hSaltKey],
+            salt: salt,
+            info: bytes_to_slice!(params.pInfo, params.ulInfoLen, u8).to_vec(),
+            emit_data_obj: mech.mechanism == CKM_HKDF_DATA,
+        })
+    }
+
+    fn fixup_template(
+        template: &[CK_ATTRIBUTE],
+        attributes: &[CK_ATTRIBUTE],
+    ) -> Option<Vec<CK_ATTRIBUTE>> {
+        /* this function only adds missing defaults, it ain't validation */
+        let mut vec: Option<Vec<CK_ATTRIBUTE>> = None;
+        for attr in attributes {
+            match template.iter().find(|a| a.type_ == attr.type_) {
+                Some(_) => (),
+                None => {
+                    if let Some(ref mut v) = vec {
+                        v.push(attr.clone());
+                    } else {
+                        let mut v = template.to_vec();
+                        v.push(attr.clone());
+                        vec = Some(v);
+                    }
+                }
+            }
+        }
+        vec
+    }
+
+    fn data_object_and_secret_size(
+        &self,
+        template: &[CK_ATTRIBUTE],
+        objfactories: &ObjectFactories,
+    ) -> KResult<(Object, usize)> {
+        let default_class = CKO_DATA;
+        let mut otmpl = Self::fixup_template(
+            template,
+            &[CK_ATTRIBUTE::from_ulong(CKA_CLASS, &default_class)],
+        );
+        let keysize = match template.iter().find(|a| a.type_ == CKA_VALUE_LEN) {
+            Some(cka) => {
+                let ks = cka.to_ulong()? as usize;
+                /* we must remove CKA_VALUE_LEN from the template as it is not a valid
+                 * attribute for a CKO_DATA object */
```

Comment:
The value is absolutely legit in the template and we use it to determine the actual requested derived key size.
But this attribute is invalid on a data object (the create function will fail if we leave it there) so we have to remove it from the template.

This is a special case because this mechanism returns a CKO_DATA object instead of a CKO_SECRET_KEY.

Do not ask me why PKCS#11 committee thought it was a good idea to break all the semantics this way when returning an extractable key would have been the seemling more logical thing to do.